File: combinatorics/resource/ST/project_5/bai_toan_7/Bottom-up.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// Dinh nghia cau truc nut
struct TreeNode {
    string label;                         // Nhan cua nut
    vector<TreeNode*> children;          // Danh sach con
    TreeNode(string l) : label(l) {}
};

// Duyet bottom-up bang de quy
void bottomupTraversal(TreeNode* node) {
    if (!node) return;
    for (TreeNode* child : node->children)
        bottomupTraversal(child);        // Duyet con truoc
    cout << node->label << endl;         // In nut sau
}

// Nhap cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan nut: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i+1 << " cua " << label << ":\n";
        node->children.push_back(inputTree());
    }
    return node;
}

// Ham main
int main() {
    cout << "Nhap cay:\n";
    TreeNode* root = inputTree();
    cout << "Duyet bottom-up:\n";
    bottomupTraversal(root);
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_7/Bottom-up.py
# Lop dinh nghia cay
class TreeNode:
    def __init__(self, label):
        self.label = label                  # Luu nhan cua nut
        self.children = []                 # Danh sach cac nut con

# Duyet bottom-up bang de quy
def bottomup_traversal(node):
    if node is None:
        return
    for child in node.children:            # Duyet tat ca con truoc
        bottomup_traversal(child)
    print(node.label)                      # Sau khi duyet con moi in nut hien tai

# Ham nhap cay
def input_tree():
    label = input("Nhap nhan nut: ")
    node = TreeNode(label)
    n = int(input(f"So con cua {label}: "))
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()
        node.children.append(child)
    return node

# Ham chinh
if __name__ == "__main__":
    print("Nhap cay:")
    root = input_tree()
    print("Duyet bottom-up:")
    bottomup_traversal(root)